test/extended/apiserver/tls.go

```diff
@@ -1,31 +1,467 @@
+// openshift/conformance/parallel test suite, and it is expected that there are jobs
+// which run that entire conformance suite against clusters running any TLS profiles
+// that there is a desire to test.
+var _ = g.Describe("[sig-api-machinery][Feature:APIServer]", func() {
+	defer g.GinkgoRecover()
+
+	oc := exutil.NewCLI(namespace)
+
+	g.It("TestTLSMinimumVersions", func() {
+		ctx := context.TODO()
+
+		isMicroShift, err := exutil.IsMicroShiftCluster(oc.AdminKubeClient())
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		isHyperShift, err := exutil.IsHypershift(ctx, oc.AdminConfigClient())
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		if isMicroShift || isHyperShift {
+			g.Skip("tls configuration for the apiserver resource is not applicable to microshift or hypershift clusters - skipping")
+		}
+
+		ipFamily := getIPFamilyForCluster(*oc, oc.Namespace())
+
+		if ipFamily != IPv4 {
+			g.Skip("tls configuration is only tested on IPv4 clusters, skipping")
+		}
+
+		insecure := true
+		configFlags := &genericclioptions.ConfigFlags{}
+		configFlags.Insecure = &insecure
+		configFlags.APIServer = &oc.AdminConfig().Host
+		configFlags.BearerToken = &oc.AdminConfig().BearerToken
+
+		config, err := oc.AdminConfigClient().ConfigV1().APIServers().Get(ctx, "cluster", metav1.GetOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		var tlsShouldWork, tlsShouldNotWork *tls.Config
+
+		if config.Spec.TLSSecurityProfile == nil {
+			// default to intermediate profile, which requires 1.2
+			tlsShouldWork = &tls.Config{MinVersion: tls.VersionTLS12, MaxVersion: tls.VersionTLS13, InsecureSkipVerify: true}
+			tlsShouldNotWork = &tls.Config{MinVersion: tls.VersionTLS11, MaxVersion: tls.VersionTLS11, InsecureSkipVerify: true}
+		} else {
+			switch config.Spec.TLSSecurityProfile.Type {
+			case configv1.TLSProfileIntermediateType:
+				tlsShouldWork = &tls.Config{MinVersion: tls.VersionTLS12, MaxVersion: tls.VersionTLS13, InsecureSkipVerify: true}
+				tlsShouldNotWork = &tls.Config{MinVersion: tls.VersionTLS11, MaxVersion: tls.VersionTLS11, InsecureSkipVerify: true}
+
+			case configv1.TLSProfileModernType:
+				tlsShouldWork = &tls.Config{MinVersion: tls.VersionTLS13, MaxVersion: tls.VersionTLS13, InsecureSkipVerify: true}
+				tlsShouldNotWork = &tls.Config{MinVersion: tls.VersionTLS12, MaxVersion: tls.VersionTLS12, InsecureSkipVerify: true}
+
+			default:
+				g.Skip("only intermediate or modern profiles are tested")
+			}
+		}
+
+		//////
+
+		g.By("Checking the Kube API server")
+
+		err = ForwardPortsAndExecute(
+			"apiserver",
+			"openshift-kube-apiserver",
+			[]string{"443"},
+			3,
+			200*time.Millisecond,
+			func(port int) {
+				conn, err := tls.Dial("tcp", fmt.Sprintf("localhost:%d", port), tlsShouldWork)
+				o.Expect(err).NotTo(o.HaveOccurred())
+
+				conn.Close()
+
+				_, err = tls.Dial("tcp", fmt.Sprintf("localhost:%d", port), tlsShouldNotWork)
+				o.Expect(err).To(o.HaveOccurred())
```

Comment:
what type/kind of err do we expect here? how are we going to distinguish it from a network err ?